netket/logging/state_log.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import dataclasses
+import tarfile
+import time
+from io import BytesIO
+import shutil
+import glob
+
+import os
+from os import path as _path
+import numpy as np
+import jax
+
+from flax import serialization
+
+
+def save_binary_to_tar(tar_file, byte_data, name):
+    abuf = BytesIO(byte_data)
+
+    # Contruct the info object with the correct length
+    info = tarfile.TarInfo(name=name)
+    info.size = len(abuf.getbuffer())
+
+    # actually save the data to the tar file
+    tar_file.addfile(tarinfo=info, fileobj=abuf)
+
+
+class StateLog:
+    """
+        Tar Logger, serializing the variables of the variational state during a run.
+
+    : bool flag indicating whever to store variables in a tar file. The tar archive will
+                    contain a file with numbers going from 0 to N, and every file corresponds to the variables of
+                    the variational state at that step.
+
+
+        Data is serialized to a tar archive as many files named `[0.mpack, 1.mpack, ...]`, where
+        the number increases by 1 every time it's called.
+
+        The tar file inside is not flushed to disk (closed) until this object is deleted or python
+        is shut down.
+    """
+
+    def __init__(
+        self,
+        output_prefix: str,
+        mode: str = "write",
+        save_every: int = 1,
+        tar: bool = False,
+    ):
+        """
+        Construct a Tar Logger.
+
+        Args:
+            output_prefix: the name of the output file before the extension (if tar=True) or of the
+                output folder.
+            save_every: every how many iterations the variables should be saved. (default 1)
+            mode: Specify the behaviour in case the file already exists at this output_prefix. Options are
+                - `[w]rite`: (default) overwrites file/delete the folder if it already exists;
+                - `[a]ppend`: appends to the file/folder if it exists, overwise creates a new file;
+                - `[x]` or `fail`: fails if file/folder already exists;
+            tar: if True creates a tar archive instead of a folder.
```

Comment:
```suggestion
            mode: Specify the behaviour in case the file already exists at this output_prefix. Options are
                - `[w]rite`: (default) overwrites file/delete the folder if it already exists;
                - `[a]ppend`: appends to the file/folder if it exists, overwise creates a new file;
                - `[x]` or `fail`: fails if file/folder already exists;
            save_every: every how many iterations the variables should be saved. (default: 1)
            tar: if True, state files are saved in a TAR archive instead of a folder. (default: False)
```